clustered/server/src/test/java/org/ehcache/clustered/server/store/ServerStoreTest.java

```diff
@@ -0,0 +1,227 @@
+      assertThat(readPayLoad(element.getPayload()), is(Long.valueOf(1)));
+    }
+  }
+
+  @Test
+  public void testAppendMappingExists() {
+    ServerStore store = newStore();
+    populateStore(store);
+    store.append(2, getPayload(22));
+    Chain chain = store.get(2);
+    assertThat(chain.isEmpty(), is(false));
+    Iterator<Element> linkIterator = chain.iterator();
+    Element element1 = linkIterator.next();
+    assertThat(readPayLoad(element1.getPayload()), is(Long.valueOf(2)));
+    Element element2 = linkIterator.next();
+    assertThat(readPayLoad(element2.getPayload()), is(Long.valueOf(22)));
+
+  }
+
+  @Test
+  public void testGetAndAppendNoMappingExists() {
+    ServerStore store = newStore();
+    Chain chain = store.getAndAppend(1, getPayload(1));
+    assertThat(chain.isEmpty(), is(true));
+    chain = store.get(1);
+    for (Element element : chain) {
+      assertThat(readPayLoad(element.getPayload()), is(Long.valueOf(1)));
+    }
+  }
+
+  @Test
+  public void testGetAndAppendMappingExists() {
+    ServerStore store = newStore();
+    populateStore(store);
+    Chain chain = store.getAndAppend(1, getPayload(22));
+    for (Element element : chain) {
+      assertThat(readPayLoad(element.getPayload()), is(Long.valueOf(1)));
+    }
+    chain = store.get(1);
+    Iterator<Element> linkIterator = chain.iterator();
+    Element element1 = linkIterator.next();
+    assertThat(readPayLoad(element1.getPayload()), is(Long.valueOf(1)));
+    Element element2 = linkIterator.next();
+    assertThat(readPayLoad(element2.getPayload()), is(Long.valueOf(22)));
+
+  }
+
+  @Test
+  public void testReplaceAtHeadSucceedsMappingExistsHeadMatchesStrictly() {
+    ServerStore store = newStore();
+    populateStore(store);
+    Chain existingMapping = store.get(1);
+
+    store.replaceAtHead(1, existingMapping, chainBuilder.build(compactChain(existingMapping, getPayload(11))));
+    Chain chain = store.get(1);
+    for (Element element : chain) {
+      assertThat(readPayLoad(element.getPayload()), is(Long.valueOf(11)));
+    }
+
+    store.append(2, getPayload(22));
+    store.append(2, getPayload(222));
+
+    existingMapping = store.get(2);
+
+    store.replaceAtHead(2, existingMapping, chainBuilder.build(compactChain(existingMapping, getPayload(2222))));
+
+    chain = store.get(2);
+
+    for (Element element : chain) {
+      assertThat(readPayLoad(element.getPayload()), is(Long.valueOf(2222)));
+    }
+  }
+
+  @Test
+  public void testReplaceAtHeadSucceedsMappingExistsHeadMatches() {
+    ServerStore store = newStore();
+    populateStore(store);
+
+    Chain existingMapping = store.get(1);
+
+    store.append(1, getPayload(11));
+
+    store.replaceAtHead(1, existingMapping, chainBuilder.build(compactChain(existingMapping, getPayload(111))));
+    Chain chain = store.get(1);
+    Iterator<Element> elements = chain.iterator();
+
+    assertThat(readPayLoad(elements.next().getPayload()), is(Long.valueOf(111)));
```

Comment:
Isn't this wrong?  The list before the replaceAtHead was: [1, 11], and the replace was ([1], [111]), so the result should be [111, 11], right?  So when iterator in reverse order you'll see [11] then [111].  Or am I getting confused?